### ISSUE TITLE
SIGABRT: Assertion failed: (ec == 0), function unlock, file /BuildRoot/Library/Caches/com.apple.xbs/Sources/libcxx/libcxx-400.9.4/src/mutex.cpp, line 48

### ISSUE THREAD

**Issue by dennisklein** [FairRootGroup/DDS#276]
**DDS**: `3.0-2-g8b00716`
**OS**: macOS `10.14.6 (18G1012)`, Darwin `18.7.0`
**Compiler**: AppleClang `10.0.1 (clang-1001.0.46.4)`
**Boost**: `1.68.0`
**Host**: witmac007.gsi.de (in @MohammadAlTurany's office)
**Failed test**: https://cdash.gsi.de/testDetails.php?test=6102012&build=221938
**DDS log**: n.a.
**Summary**: https://cdash.gsi.de/buildSummary.php?buildid=221938

**Error**:
```
Assertion failed: (ec == 0), function unlock, file /BuildRoot/Library/Caches/com.apple.xbs/Sources/libcxx/libcxx-400.9.4/src/mutex.cpp, line 48.
/Users/alfaci/alfaci/workspace/FairRootGroup_FairMQ_dev/build/examples/dds/fairmq-start-ex-dds.sh: line 76: 64098 Abort trap: 6           dds-info --active-topology
```

**Commands to run the failing test** (system installations of ZeroMQ or Boost should be found without any argument):
```
$ git clone -b v1.4.11 https://github.com/FairRootGroup/FairMQ && cd FairMQ
FairMQ/ $ cmake -S. -Bbuild -DDDS_ROOT=/path/to/dds -DFairLogger_ROOT=/path/to/fairlogger -DBoost_ROOT=/path/to/boost -DZeroMQ_ROOT=/path/to/zeromq -DBUILD_DDS_PLUGIN=ON -DBUILD_SDK=ON -DBUILD_SDK_COMMANDS=ON -DCMAKE_INSTALL_PREFIX=./install
FairMQ/ $ cmake --build build --target install -j8
FairMQ/ $ cd build
FairMQ/build/ $ ctest -R "Example.DDS.localhost" -VV
or
FairMQ/build/ $ examples/dds/fairmq-start-ex-dds.sh localhost
```

Test was failing when all tests were run with `ctest -j12`.

**Comment by dennisklein** [comment]
Another variant: https://cdash.gsi.de/testDetails.php?test=6206874&build=224317

**Comment by AnarManafov** [comment]
@dennisklein , wanted to ping you about this issue.
Do you still observe it with the latest DDS master?

I was not able to reproduce this stack overflow error.

**Comment by dennisklein** [comment]
Sry for the late reply. I checked with latest DDS 3.5.1 (and Boost 1.71 on macOS 10.14.6 on `witmac007.gsi.de`) and the error is still happening in about 15% of the invocations.

```
alfaci at witmac007 in ~/dklein/FairMQ/build on master*$ ctest -R Example.DDS.localhost -VV
UpdateCTestConfiguration  from :/Users/alfaci/dklein/FairMQ/build/DartConfiguration.tcl
Parse Config file:/Users/alfaci/dklein/FairMQ/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/Users/alfaci/dklein/FairMQ/build/DartConfiguration.tcl
Parse Config file:/Users/alfaci/dklein/FairMQ/build/DartConfiguration.tcl
Test project /Users/alfaci/dklein/FairMQ/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 177
    Start 177: Example.DDS.localhost

177: Test command: /Users/alfaci/dklein/FairMQ/build/examples/dds/fairmq-start-ex-dds.sh "localhost"
177: Test timeout computed to be: 15
177: DDS session ID: 6f66d85a-f5b3-474a-bb43-6fcd0d0548f1
177: Checking precompiled binaries for the local system only:
177: 	dds-wrk-bin-3.5.1-Darwin-universal.tar.gz - OK
177: Starting DDS commander...
177: Waiting for DDS Commander to appear online...
177: DDS commander appears online. Testing connection...
177: DDS commander is up and running.
177: ------------------------
177: DDS commander server: 46256
177: ------------------------
177: Startup time: 53.0317 ms
177: Default DDS session is set to 6f66d85a-f5b3-474a-bb43-6fcd0d0548f1
177: Currently running DDS sessions:
177:    3fe24218-c3ca-4e98-878b-9a5ffa55496c 	 [2020-07-21T09:57:13Z] 	 RUNNING
177:    a33708df-ff1a-48b4-b9f0-7874b1417643 	 [2020-07-21T09:57:28Z] 	 RUNNING
177:  * 6f66d85a-f5b3-474a-bb43-6fcd0d0548f1 	 [2020-07-21T10:00:24Z] 	 RUNNING
177: SESSION ID: 6f66d85a-f5b3-474a-bb43-6fcd0d0548f1
177: dds-submit: Connection established.
177: dds-submit: Requesting server to process job submission...
177: dds-submit: Server reports: Creating new worker package...
177: dds-submit: Server reports: RMS plug-in: /Users/alfaci/dklein/plugins/dds-submit-localhost/dds-submit-localhost
177: dds-submit: Server reports: Initializing RMS plug-in...
177: dds-submit: Server reports: RMS plug-in is online. Startup time: 17ms.
177: dds-submit: Server reports: Plug-in: Will use the local host to deploy 1 agents
177: dds-submit: Server reports: Plug-in: Using '/var/folders/jq/cspf26md6b7crchgvwyy_3fc0000gp/T/6f66d85a-f5b3-474a-bb43-6fcd0d0548f1/dds_2020-07-21-10-00-25-513' to spawn agents
177: dds-submit: Server reports: Plug-in: Starting DDSScout in '/var/folders/jq/cspf26md6b7crchgvwyy_3fc0000gp/T/6f66d85a-f5b3-474a-bb43-6fcd0d0548f1/dds_2020-07-21-10-00-25-513/wn'
177: dds-submit: Server reports: Plug-in: DDS agents have been submitted
177: dds-submit: Server reports: Plug-in: Checking status of agents...
177: dds-submit: Server reports: Plug-in: All agents have been started successfully
177: dds-submit: Server reports: Plug-in: Validating...
177: dds-submit: Server reports: Plug-in: All agents have been validated successfully
177: dds-submit: Submission took: 507.268 ms
177: ...waiting for 12 idle slots...
177: Active agents online: 12
177: Idle agents online: 12
177: Executing agents online: 0
177: TOPOLOGY FILE: /Users/alfaci/dklein/FairMQ/build/examples/dds/ex-dds-topology.xml
177: TOPOLOGY NAME: ExampleDDS
177: no active topology
177: dds-topology: Connection established.
177: dds-topology: Requesting server to ACTIVATE a topology...
177: [==================================================] 100 % (1/1)
177: Updated agent topologies: 1
177: Errors: 0
177: Total: 1
177: Time to update agent topologies: 0.001 s
[==================================================] 100 % (12/12)2)
177: Assigned/Uploaded tasks: 12
177: Errors: 0
177: Total: 12
177: Time to assign/upload: 0.004 s
[==================================================] 100 % (12/12)2)
177: Assigned/Uploaded tasks: 12
177: Errors: 0
177: Total: 12
177: Time to assign/upload: 0.075 s
[==================================================] 100 % (12/12)2)
177: Activated tasks: 12
177: Errors: 0
177: Total: 12
177: Time to Activate: 0.025 s
177: active topology: ExampleDDS; path: /Users/alfaci/dklein/FairMQ/build/examples/dds/ex-dds-topology.xml
177: ...waiting for 12 executing slots...
177: Active agents online: 12
177: Idle agents online: 0
177: Executing agents online: 12
177: ------------------------
177: ...waiting for Topology to finish...
177: > waiting for all to reach IDLE
177: > initiating InitDevice transition --> all
177: > initiating CompleteInit transition --> all
177: > initiating Bind transition --> all
177: > initiating Connect transition --> all
177: > initiating InitTask transition --> all
177: > initiating Run transition --> all
177: > waiting for main/(Sampler|Sink).* to reach RUNNING->READY
177: ...main/(Sampler|Sink).* are READY, sending shutdown...
177: > initiating Stop transition --> all
177: [ERROR] STOP transition failed for main/Sink_0
177: [ERROR] STOP transition failed for main/Sampler_0
177: > initiating ResetTask transition --> all
177: > initiating ResetDevice transition --> all
177: > initiating End transition --> all
177: ...waiting for 12 idle slots...
177: Assertion failed: (ec == 0), function unlock, file /BuildRoot/Library/Caches/com.apple.xbs/Sources/libcxx/libcxx-400.9.4/src/mutex.cpp, line 48.
177: Active agents online: 12
177: Idle agents online: 12
177: Executing agents online: 0
177: /Users/alfaci/dklein/FairMQ/build/examples/dds/fairmq-start-ex-dds.sh: line 70: 46560 Abort trap: 6           dds-info --idle-count --wait ${requiredNofSlots}
177: Stopping DDS commander: 6f66d85a-f5b3-474a-bb43-6fcd0d0548f1
177: Sending a graceful stop signal to Commander (pid/sessionID): 46256/6f66d85a-f5b3-474a-bb43-6fcd0d0548f1
177: dds-commander: self exiting (46256)...
177: .
177: CLEANUP PERFORMED
1/1 Test #177: Example.DDS.localhost ............***Failed  Required regular expression not found. Regex=[Example successful
]  7.01 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   7.03 sec

The following tests FAILED:
	177 - Example.DDS.localhost (Failed)
Errors while running CTest
```

In another run it failed in a different `dds-info` invocation:

```
177: Assertion failed: (ec == 0), function unlock, file /BuildRoot/Library/Caches/com.apple.xbs/Sources/libcxx/libcxx-400.9.4/src/mutex.cpp, line 48.
177: active topology: ExampleDDS; path: /Users/alfaci/dklein/FairMQ/build/examples/dds/ex-dds-topology.xml
177: /Users/alfaci/dklein/FairMQ/build/examples/dds/fairmq-start-ex-dds.sh: line 48: 51717 Abort trap: 6           dds-info --active-topology
```

You can find the test script from above in https://github.com/FairRootGroup/FairMQ/blob/v1.4.21/examples/dds/fairmq-start-ex-dds.sh.in (accompanying files are in the same directory).

Logs for the above sessions:
* `6f66d85a-f5b3-474a-bb43-6fcd0d0548f1`: [dds_2020-07-21.0.log](https://github.com/FairRootGroup/DDS/files/4952152/dds_2020-07-21.0.log), [wn_6f66d85a-f5b3-474a-bb43-6fcd0d0548f1.tar.gz](https://github.com/FairRootGroup/DDS/files/4952171/wn_6f66d85a-f5b3-474a-bb43-6fcd0d0548f1.tar.gz)
* `ed71f1d7-a790-4c31-8061-5d9356f46721`: [dds_2020-07-21.0.log](https://github.com/FairRootGroup/DDS/files/4952179/dds_2020-07-21.0.log), [wn_ed71f1d7-a790-4c31-8061-5d9356f46721.tar.gz](https://github.com/FairRootGroup/DDS/files/4952175/wn_ed71f1d7-a790-4c31-8061-5d9356f46721.tar.gz)

**Comment by rbx** [comment]
For the record - it looks like the issue only appears on a particular system in our test setup - the witmac007 (`MacOS10.14-x86_64-AppleLLVM10.0.1`). There it occurs consistently.
The other MacOS machine - demac016 (MacOS10.13-x86_64-AppleLLVM10.0.0) never has this issue. @dennisklein correct me if I'm wrong).

Perhaps you could run your DDS test suite and/or try to reproduce it there (witmac007), or on a similar system.

**Comment by dennisklein** [comment]
> Perhaps you could run your DDS test suite and/or try to reproduce it there (witmac007), or on a similar system.

Will try on another system. FYI, for the most recent test, I used the apple compiler installation from witmac007 as before, but used a freshly installed Boost, DDS, and FairMQ.

**Comment by rbx** [comment]
My comment was actually more for @AnarManafov , in case he cannot reproduce this in his environment, but we can also try another machine, if available.

**Comment by AnarManafov** [comment]
@rbx , @dennisklein , guys, thanks for the info.
If possible, please try on another system. Any additional data will help.
On my local test, I think, I caught it once (at least something similar), but can't reproduce it afterwards.

**Comment by AnarManafov** [comment]
@dennisklein , do you run these tests in parallel or sequentially?

**Comment by dennisklein** [comment]
> @dennisklein , do you run these tests in parallel or sequentially?

Sequentially: https://github.com/FairRootGroup/FairMQ/blob/5e97d8595659edeb87ea6aa47a64340e3c9f2ed8/examples/dds/CMakeLists.txt#L45

**Comment by dennisklein** [comment]
Some update:

Host: `witmac007`
Clang: `11.0 build 1100`
macOS: `10.14.6-x86_64`
CLT: `11.3.1.0.1.1576735732`
Xcode: `11.3.1`

Even after major Xcode (from 10 to 11) update we are hitting this bug: https://cdash.gsi.de/testDetails.php?test=7391051&build=252179

Googling this issue reveals that also other popular programs are affected: [Apache Arrow](http://mail-archives.apache.org/mod_mbox/arrow-issues/201910.mbox/%3CJIRA.13264082.1571852811000.28881.1571869620309@Atlassian.JIRA%3E), [Adobe InDesign](https://community.adobe.com/t5/indesign/indesign-2020-and-2021-crash-on-quit-with-typequest-catalogparser-thread/td-p/11548708?page=1), [Cubase](https://www.steinberg.net/forums/viewtopic.php?t=185693).

Good news is that it appears to be limited to macOS 10.14, we have not seen it on 10.15.

**Comment by AnarManafov** [comment]
Hey @dennisklein,

if I understand the background of this bug correctly it has something to do with an orphan mutex left in a locked state. For example, when an object has a method which locks the mutex, but also deletes itself (the object) leaving the mutex abandoned in a locked state. It could also happen with lambdas using a "this" pointer with the same mutex exercise.

I tried my best and literally checked every place where we use locks for even a theoretical possibility for this bug to happen. Unfortunately I didn't find anything like that.

I strongly suspect it has something to do with this given version of OS X. For instance, with STD implementation provided with this version of OS X.

I would propose to close this issue (put on hold) until we face it on linux (our production target) or newer versions of OS X.
Please, feel free to reopen if you think we still have to try to find the source of the problem.
